### ISSUE TITLE
TS2339: Property 'bar' does not exist on type 'never' (strictNullChecks)

### ISSUE THREAD

**Issue by maksimr** [microsoft/TypeScript#41766]
Example code:

index.js
```javascript
const [foo] = [null];
console.log(foo?.bar);
```

tsconfig.json
```json
{
  "compilerOptions": {
    "target": "es6",
    "allowJs": true,
    "checkJs": true,
    "noEmit": true,
    "strictNullChecks": true,
    "moduleResolution": "node",
    "resolveJsonModule": true
  }
}
```

Command:
```bash
tsc
```

AR:
```
error TS2339: Property 'bar' does not exist on type 'never'.

2 console.log(foo?.bar);
                   ~~~
Found 1 error.
```

ER: No error

------
Typescript version 4.1.2

If change  the code on:
```javascript
const bar = [null];
const foo = bar[0];
console.log(foo?.bar);
```

All works fine

**Comment by RyanCavanaugh** [comment]
By what justification should this not be an error? This code doesn't make any sense.

**Comment by maksimr** [comment]
@RyanCavanaugh this is just a minimal reproducible example without any attached business logic to the code. The issue that typescript throws error when we handle null value.
Code like this
```
const bar = [null];
const foo = bar[0];
```
does not throw error.

In real world use case we use `useState` with initial value `null` and after that we set an object. We do the same null check but typescript throw error and this happens only when destraction is used

**Comment by RyanCavanaugh** [comment]
Seems like you're talking about #9998 then

**Comment by maksimr** [comment]
@RyanCavanaugh kind of... mostly about optional chaining `foo?.bar` and strict null check. What I want is eliminate `Property '<name>' does not exist on type 'never'.` error without disabling strict null check. For me in the code above chaining operator should be enough to say typescript analyzer that I know about possible null value and cover it by optional chaining operator `?` so all null safe :)
